Add documentation on installation and usage (#132)
open-telemetry/opentelemetry-dotnet-instrumentation: docs/USAGE.md

# Usage

**WARNING**: Please notice that no official release has been created for this repo yet, so installation instructions below would require you to build the code manually beforehand.

## Configure the OpenTelemetry Tracing Library for .NET

### Configuration values

> Use these environment variables to configure the tracing library:

Reviewer: Missing documentation for:
- OTEL_TRACE_CONFIG_FILE
- OTEL_VERSION
- ...

Please double-check everything with: https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/blob/main/src/Datadog.Trace/Configuration/ConfigurationKeys.cs 😉

> | Environment variable | Default value | Description |

Reviewer: Can we change the table formatting to the one used here: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/sdk-environment-variables.md to be consistent?

|-|-|-|

> | `OTEL_SERVICE_NAME` |  | The name of the service. |

Reviewer: Should it not be `OTEL_SERVICE` ?

Reviewer 2: I tend to prefer `OTEL_SERVICE_NAME` - eventually this should come from resources, anyway, I think that the suffix `_NAME` makes it a bit more intuitive. Happy to be convinced otherwise.

Author: I don't what to change that in this PR as this is purely documentation issue. Please create GH issue if you feel the other name would be more fitting.

> | `OTEL_ENV` |  | The value for the `environment` tag added to every span. This determines the environment in which the service is available in OTEL µAPM.  |

Reviewer 2: IIRC `environment` is an OpenTracing convention, we could actually start to use this one https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/semantic_conventions/deployment_environment.md

```suggestion
| `OTEL_ENV` |  | The value for the `environment` tag added to every span. |
```

| `OTEL_TRACE_ENABLED` | `true` | Enable to activate the tracer. |
| `OTEL_TRACE_DEBUG` | `false` | Enable to activate debugging mode for the tracer. |

> | `OTEL_TRACE_AGENT_URL` | `http://localhost:9080/v1/trace` | The hostname and port for a OTEL Smart Agent or OpenTelemetry Collector. |

Reviewer 3: The example value and description do not match. One has a full URL and the other says that just a hostname and port should be provided. Which one is correct?

> | `OTEL_TRACE_AGENT_URL` | `http://localhost:9080/v1/trace` | The hostname and port for a OTEL Smart Agent or OpenTelemetry Collector. |

Reviewer 2: For the time being:
```suggestion
| `OTEL_TRACE_AGENT_URL` | `http://localhost:9080/v1/trace` | The URL to where send the traces. |
```
Could you please open an issue to use env vars from the exporters instead?

| `OTEL_TRACE_GLOBAL_TAGS` |  | Comma-separated list of key-value pairs to specify global span tags. For example: `"key1:val1,key2:val2"` |
| `OTEL_LOGS_INJECTION` | `false` | Enable to inject trace IDs, span IDs, service name and environment into logs. This requires a compatible logger or manual configuration. |
| `OTEL_MAX_LOGFILE_SIZE` | `10 MB` | The maximum size for tracer log files, in bytes. |
| `OTEL_TRACE_LOG_PATH` | Linux: `/var/log/OTEL/dotnet/dotnet-profiler.log`<br>Windows: `%ProgramData%"\OTEL .NET Tracing\logs\dotnet-profiler.log` | The path of the profiler log file. |
| `OTEL_DIAGNOSTIC_SOURCE_ENABLED` | `true` | Enable to generate troubleshooting logs with the `System.Diagnostics.DiagnosticSource` class. |

> | `OTEL_DISABLED_INTEGRATIONS` |  | The integrations you want to disable, if any, separated by a semi-colon. These are the supported integrations: AspNetMvc, AspNetWebApi2, DbCommand, ElasticsearchNet5, ElasticsearchNet6, GraphQL, HttpMessageHandler, IDbCommand, MongoDb, NpgsqlCommand, OpenTracing, ServiceStackRedis, SqlCommand, StackExchangeRedis, Wcf, WebRequest |

Reviewer 2: Could you please double-check if it is really `;` ? I remember being mentioned in some other PR that we should standardized to `,`. If currently is semi-colon please open an issue to fix it.

Author: Yes, this uses `;`:
`internal static class ConfigurationSourceExtensions
    {
        private const char Separator = ';';`

Reviewer: @RassK FYI

Reviewer 2: @dszmigielski resolving the review, please do not forget to create the issue.

> | `OTEL_TRACE_DOMAIN_NEUTRAL_INSTRUMENTATION` | `false` |  Sets whether to intercept method calls when the caller method is inside a domain-neutral assembly. This is recommended when instrumenting IIS applications. |

Reviewer 2: I don't think this one exists anymore.

Author: It's present in environment_variables.h and is used further in some cpp code.

Reviewer 2: Thanks for pointing that out.

> | `OTEL_PROFILER_PROCESSES` |  | Sets the filename of executables the profiler can attach to. If not defined (default), the profiler will attach to any process. Supports multiple values separated with semi-colons, for example: `MyApp.exe;dotnet.exe` |

Reviewer 2: Same regarding previous env var using `;` as a separator - not to be fixed on this PR but please take the chance to access the current state of such env vars. Same for next env var.

Author: This is accurate, they still use `;`

| `OTEL_PROFILER_EXCLUDE_PROCESSES` |  | Sets the filename of executables the profiler cannot attach to. If not defined (default), the profiler will attach to any process. Supports multiple values separated with semi-colons, for example: `MyApp.exe;dotnet.exe` |

> | `OTEL_CONVENTION` | `Datadog` | Sets the outbound http and trace id convention for tracer. Available values are: `Datadog` (64bit trace id), `OpenTelemetry` (128 bit trace id). |

Reviewer 2: also the semantic convention for tracing

> | `OTEL_PROPAGATOR` | `Datadog` | Sets the propagator for tracer. Available values are: `Datadog`, `B3`, `W3C`. |

Reviewer 2: We can use separators here, right? The order is also relevant, correct?

### Linux

After downloading the library, install the CLR Profiler and its components
via your system's package manager.

1. Download the latest release of the library.
2. Install the CLR Profiler and its components with your system's package
manager:
    ```bash

>     # Use dpkg:

Reviewer 2: Sometimes people are using docker containers and they are not sure about which distro their Linux is based on. Perhaps add some not along the lines of https://github.com/signalfx/signalfx-dotnet-tracing#linux-instrumentation-not-working

Reviewer 2: Ah I see it below, perhaps add a link on item # 2

    $ dpkg -i otel-dotnet-autoinstrumentation.deb

    # Use rpm:
    $ rpm -ivh otel-dotnet-autoinstrumentation.rpm

    # Install directly from the release bundle:
    $ tar -xf otel-dotnet-autoinstrumentation.tar.gz -C /

    # Install directly from the release bundle for musl-using systems (Alpine Linux):
    $ tar -xf otel-dotnet-autoinstrumentation-musl.tar.gz -C /
    ```
1. Configure the required environment variables:
    ```bash
    $ source /opt/otel-dotnet-autoinstrumentation/defaults.env
    ```
2. Set the service name:
    ```bash
    $ export OTEL_SERVICE_NAME='MyCoreService'
    ```
3. Set Zipkin exporter:
    ```bash
    $ export OTEL_EXPORTER='Zipkin'
    ```
4. Set OpenTelemetry conventions:
    ```bash
    $ export OTEL_CONVENTION='OpenTelemetry'
    ```

> 5. Set the endpoint of a Smart Agent or OpenTelemetry Collector:

Reviewer 2: Do not use Smart Agent. In this case something like "Set the endpoint, e.g. OpenTelemetry Collector:"

    ```bash

>     $ export OTEL_TRACE_AGENT_URL='http://<YourSmartAgentOrCollector>:9080/v1/trace'

Reviewer 2: ```suggestion
    $ export OTEL_TRACE_AGENT_URL='http://localhost:9080/v1/trace'
```

    ```
7. Optionally, create the default logging directory:
    ```bash
    $ source /opt/otel-dotnet-autoinstrumentation/createLogPath.sh
    ```
8. Run your application:
    ```bash
    $ dotnet run
    ```

### Windows

**Warning**: Pay close attention to the scope of environment variables. Ensure they
are properly set prior to launching the targeted process. The following steps set the
environment variables at the machine level, with the exception of the variables used
for finer control of which processes will be instrumented, which are set in the current
command session.

1. Install the CLR Profiler using an installer file (`.msi` file) from the latest release.
Choose the installer (x64 or x86) according to the architecture of the application
you're instrumenting.
2. Configure the required environment variables to enable the CLR Profiler:
    - For .NET Framework applications:
    ```batch
    setx COR_PROFILER "{918728DD-259F-4A6A-AC2B-B85E1B658318}" /m
    ```
   - For .NET Core applications:
   ```batch
   setx CORECLR_PROFILER "{918728DD-259F-4A6A-AC2B-B85E1B658318}" /m
   ```
3. Set the "service name" that better describes your application:
   ```batch

>    setx OTEL_SERVICE_NAME MyServiceName /m

Reviewer: I guess it should be `OTEL_SERVICE`

   ```

> 4. Set the endpoint of a Smart Agent or OpenTelemetry Collector that will forward

Reviewer 2: See previous comment

the trace data:
   ```batch
   setx OTEL_TRACE_AGENT_URL http://localhost:9080/v1/trace /m
   ```
5. Optionally, enable trace injection in logs:
   ```batch
   setx OTEL_LOGS_INJECTION true /m
   ```

> 6. Optionally, if instrumenting IIS applications add the following environmet variable set to `true`:

Reviewer 2: IIRC does not apply anymore.

automatically become child spans of any existing spans in the same context:
    ```csharp
    using OpenTracing;
    using OpenTracing.Util;

    namespace MyProject
    {
        public class MyClass
        {
            public static async void MyMethod()
            {
                // Obtain the automatically registered OpenTracing.Util.GlobalTracer instance
                var tracer = GlobalTracer.Instance;

                // Create an active span that will be automatically parented by any existing span in this context
                using (IScope scope = tracer.BuildSpan("MyTracedFunctionality").StartActive(finishSpanOnDispose: true))
                {
                    var span = scope.Span;
                    span.SetTag("MyImportantTag", "MyImportantValue");
                    span.Log("My Important Log Statement");

                    var ret = await MyAppFunctionality();

                    span.SetTag("FunctionalityReturned", ret.ToString());
                }
            }
        }
    }
    ```

## Troubleshooting

Check if you are not hitting one of the issues listed below.

### IIS applications not instrumenting expected services

> Set the environment variable `OTEL_TRACE_DOMAIN_NEUTRAL_INSTRUMENTATION` to `true` - without it

Reviewer 2: As mentioned above I don;t think this applies anymore.

the CLR profiler can't instrument many libraries/frameworks under IIS.

### Linux instrumentation not working

The proper binary needs to be selected when deploying to Linux, eg.: the default Microsoft .NET images are
based on Debian and should use the `deb` package, see the [Linux](#Linux) setup section.

If you are not sure what is the Linux distribution being used try the following commands:
```terminal
$ lsb_release -a
$ cat /etc/*release
$ cat /etc/issue*
$ cat /proc/version
```

### High CPU usage

The default installation of auto-instrumentation enables tracing all .NET processes on the box.
In the typical scenarios (dedicated VMs or containers), this is not a problem.
Use the environment variables `OTEL_PROFILER_EXCLUDE_PROCESSES` and `OTEL_PROFILER_PROCESSES`
to include/exclude applications from the tracing auto-instrumentation.

> These are ";" delimited lists that control the inclusion/exclusion of processes.

Reviewer 2: Update here if needed regarding the actual separator.

### Custom instrumentation not being captured

> If the code accessing `GlobalTracer.Instance` executes before any auto-instrumentation is injected

Reviewer 2: We need to double-check if this applies here too.

Reviewer: It "applies" in the sense that we still give the possibility to the instrument using OpenTracing.

However, based on https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/129#discussion_r616875205 I think this section is not needed

@zacharycmontoya could you confirm, please?

Reviewer 2: > It "applies" in the sense that we still give the possibility to the instrument using OpenTracing.

yeah, I meant in the sense that I'm not sure if this is still an issue or not.

Author: I'll remove that

    ```
2. Set the service name:
    ```bash
    $ export OTEL_SERVICE='MyCoreService'
    ```
3. Set Zipkin exporter:
    ```bash
    $ export OTEL_EXPORTER='Zipkin'
    ```
4. Set OpenTelemetry conventions:
    ```bash
    $ export OTEL_CONVENTION='OpenTelemetry'
    ```
5. Set the endpoint, e.g. OpenTelemetry Collector:
    ```bash
    $ export OTEL_TRACE_AGENT_URL='http://<YourCollector>:9080/v1/trace'
    ```
6. Optionally, enable trace injection in logs:
    ```bash
    $ export OTEL_LOGS_INJECTION=true
    ```
7. Optionally, create the default logging directory:
    ```bash
    $ source /opt/otel-dotnet-autoinstrumentation/createLogPath.sh
    ```
8. Run your application:
    ```bash
    $ dotnet run
    ```

### Windows

**Warning**: Pay close attention to the scope of environment variables. Ensure they
are properly set prior to launching the targeted process. The following steps set the
environment variables at the machine level, with the exception of the variables used
for finer control of which processes will be instrumented, which are set in the current
command session.

1. Install the CLR Profiler using an installer file (`.msi` file) from the latest release.

> Choose the installer (x64 or x86) according to the architecture of the application
> you're instrumenting.

Reviewer 4: ```suggestion
Choose the installer (x64 or x86) according to the architecture of the operating
system where it will be running.
```

Reviewer 4: This is because the x64 MSI will install both the 32-bit and 64-bit profiler onto the machine. The x86 MSI will only install the 32-bit profiler onto the machine, since it can only run 32-bit processes.

| `OTEL_TRACE_PARTIAL_FLUSH_MIN_SPANS` | The minimum number of closed spans in a trace before it's partially flushed. `OTEL_TRACE_PARTIAL_FLUSH_ENABLED` has to be enabled for this to take effect. | `500` |
| `OTEL_SERVICE` | Application's default service name. |  |
| `OTEL_ENV` | The value for the `environment` tag added to every span. |  |
| `OTEL_TRACE_ENABLED` | Enable to activate the tracer. | `true` | 
| `OTEL_TRACE_DEBUG` | Enable to activate debugging mode for the tracer. | `false` | 
| `OTEL_TRACE_AGENT_URL` | The URL to where send the traces for `Zipkin` and `DatadogAgent` exporters (see: `OTEL_EXPORTER`). | `http://localhost:8126` | 
| `OTEL_TAGS` | Comma-separated list of key-value pairs to specify global span tags. For example: `"key1:val1,key2:val2"` |  |
| `OTEL_LOGS_INJECTION` | Enable to inject trace IDs, span IDs, service name and environment into logs. This requires a compatible logger or manual configuration. | `false` | `OTEL_EXPORTER` | The exporter to be used. The Tracer uses it to encode and dispatch traces. Available values are: `DatadogAgent`, `Zipkin`, `Jeager`. | `DatadogAgent` |

> | ` 

Reviewer 3: Looks like we're adding an extra blank row in the table.

Author: Thanks for the catch! Fixed :)

| `OTEL_TRACE_TRANSPORT` | Overrides the transport to use for communicating with the trace agent. Available values are: `datagod-tcp`, `datadog-named-pipes`. | `null` |
| `OTEL_HTTP_SERVER_ERROR_STATUSES` | The application's server http statuses to set spans as errors by. | `500-599` |
| `OTEL_HTTP_CLIENT_ERROR_STATUSES` | The application's client http statuses to set spans as errors by. | `400-499` |
| `OTEL_EXPORTER_JAEGER_AGENT_HOST` | Hostname for the Jaeger agent. | `localhost` |
| `OTEL_EXPORTER_JAEGER_AGENT_PORT` | Port for the Jaeger agent. | `6831` |
| `OTEL_TRACE_PARTIAL_FLUSH_ENABLED` | Enable to activate sending partial traces to the agent. | `false` |
| `OTEL_TRACE_PARTIAL_FLUSH_MIN_SPANS` | The minimum number of closed spans in a trace before it's partially flushed. `OTEL_TRACE_PARTIAL_FLUSH_ENABLED` has to be enabled for this to take effect. | `500` |
| `OTEL_SERVICE` | Application's default service name. |  |
| `OTEL_ENV` | The value for the `environment` tag added to every span. |  |
| `OTEL_TRACE_ENABLED` | Enable to activate the tracer. | `true` | 
| `OTEL_TRACE_DEBUG` | Enable to activate debugging mode for the tracer. | `false` | 

> | `OTEL_TRACE_AGENT_URL` | The URL to where send the traces. | `http://localhost:9080/v1/trace` | 

Reviewer 2: Isn't the default dependent on the OTel exporter?

Author: Well it seems it's not, at least not in this env variable, but this is still incorrect. It should be `http://localhost:8126` following this:

`var agentUri = source?.GetString(ConfigurationKeys.AgentUri) ??
                           // default value
                           $"http://{agentHost}:{agentPort}";`

if the agnetHost and agentPort are not set in env variables then it takes constant defaults: `localhost` and `8126` for port.

Jaeger uses its own configuration key (`OTEL_EXPORTER_JAEGER_AGENT_HOST`). Zipkin and default one use `OTEL_TRACE_AGENT_URL`. I don't know if it should work that way.

Reviewer 2: We need to fix Zipkin to also use its own env var. Now that config sources are available that shouldn't be a problem. Could you please open an issue for that?

| `OTEL_LOGS_INJECTION` | Enable to inject trace IDs, span IDs, service name and environment into logs. This requires a compatible logger or manual configuration. | `false` | `OTEL_EXPORTER` | The exporter to be used. The Tracer uses it to encode and dispatch traces. Available values are: `DatadogAgent`, `Zipkin`, `Jeager`. | `DatadogAgent` |
| ` 
| `OTEL_MAX_LOGFILE_SIZE` | The maximum size for tracer log files, in bytes. | `10 MB` |
| `OTEL_TRACE_LOG_PATH` | The path of the profiler log file. | Linux: `/var/log/OTEL/dotnet/dotnet-profiler.log`<br>Windows: `%ProgramData%"\OTEL .NET Tracing\logs\dotnet-profiler.log` |
| `OTEL_DIAGNOSTIC_SOURCE_ENABLED` | Enable to generate troubleshooting logs with the `System.Diagnostics.DiagnosticSource` class. | `true` |
| `OTEL_DISABLED_INTEGRATIONS` | The integrations you want to disable, if any, separated by a semi-colon. These are the supported integrations: AspNetMvc, AspNetWebApi2, DbCommand, ElasticsearchNet5, ElasticsearchNet6, GraphQL, HttpMessageHandler, IDbCommand, MongoDb, NpgsqlCommand, OpenTracing, ServiceStackRedis, SqlCommand, StackExchangeRedis, Wcf, WebRequest |  |

> | `OTEL_CONVENTION` | Sets the outbound http and trace id convention for tracer. Available values are: `Datadog` (64bit trace id), `OpenTelemetry` (128 bit trace id). |  `Datadog` |

Reviewer 2: I know that the OTEL conventions are not fully implemented but then we don't need to review the doc as we add each OTEL convention.
```suggestion
| `OTEL_CONVENTION` | Sets the semantic and trace id conventions for the tracer. Available values are: `Datadog` (64bit trace id), `OpenTelemetry` (128 bit trace id). |  `Datadog` |
```

> | `OTEL_PROPAGATORS` | Semicollon-separated list of the propagators for tracer. Available propagators are: `Datadog`, `B3`, `W3C`. The Tracer will try to execute extraction in the passed order. | `Datadog` |

Reviewer 2: Comma, no? If not:
```suggestion
| `OTEL_PROPAGATORS` | Semicolon separated list of the propagators for the tracer. Available propagators are: `Datadog`, `B3`, `W3C`. The Tracer will try to execute extraction in the given order. | `Datadog` |
```

> | `OTEL_TRACE_DOMAIN_NEUTRAL_INSTRUMENTATION` |  Sets whether to intercept method calls when the caller method is inside a domain-neutral assembly. This is recommended when instrumenting IIS applications. | `false` |

Reviewer 2: I don't think this env var exists anymore.

Author: This is same as [here](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/132#discussion_r620029086). It is there in cpp code.

Reviewer 2: The comment on the cpp code it is pretty good: "environment variable is not needed when running on .NET Framework 4.5.2 or higher, and will be ignored.". Adding something like: "only applies to .NET Framework prior to version 4.5.2" seems a good idea.

   reg add HKLM\SYSTEM\CurrentControlSet\Services\<ServiceName>\Environment /v CORECLR_ENABLE_PROFILING /d 1
   ```

Enable instrumentation for a specific user:
   - For .NET Framework applications:
   ```batch
   setx /s %COMPUTERNAME% /u <[domain/]user> COR_ENABLE_PROFILING 1
   ```
   - For .NET Core applications:
   ```batch
   setx /s %COMPUTERNAME% /u <[domain/]user> CORECLR_ENABLE_PROFILING 1
   ```

## Configure custom instrumentation

You can build upon the provided tracing functionality by modifying and adding
to automatically generated traces. The OpenTelemetry Tracing library for .NET
provides and registers an [OpenTracing-compatible](https://github.com/opentracing/opentracing-csharp)
global tracer you can use.

OpenTracing versions 0.11.0+ are supported and the provided tracer offers a
complete implementation of the OpenTracing API.

The auto-instrumentation provides a base you can build on by adding your own
custom instrumentation. By using both instrumentation approaches, you'll be
able to present a more detailed representation of the logic and functionality
of your application, clients, and framework.

1. Add the OpenTracing dependency to your project:
    ```xml

>     <PackageReference Include="OpenTracing" Version="0.12.0" />

Reviewer: ```suggestion
    <PackageReference Include="OpenTracing" Version="0.12.0" />
```

there is a newer version

Reviewer 2: @dszmigielski I see version 0.12.0 being used by Datadog.Trace.OpenTracing.csproj, it shouldn't be a problem if we use 0.12.1 as suggested by @pellared but it is good to confirm. To keep the PR scope: create an issue to validate the version issue.
